### PR TITLE
`files.Block`: fixed couple of bugs, added `try_prevent_shell_expansion` and support for `content: list[str]`

### PR DESCRIPTION
**Pull request by stone-w4tch3r**
https://github.com/pyinfra-dev/pyinfra/issues/1028
Removed dollar signs, looks like they were typo
https://github.com/pyinfra-dev/pyinfra/issues/984
As https://github.com/pyinfra-dev/pyinfra/issues/984#issuecomment-1565657899, renamed `files.marked_block` in documentation to `files.block`
https://github.com/pyinfra-dev/pyinfra/issues/1031
Adjusted `awk` to prevent duplications
https://github.com/pyinfra-dev/pyinfra/issues/1029
Added `try_prevent_shell_expansion` parameter to partially help with mentioned issue

And also changed type for `content` to` str | list[str]`. This makes it more unified with `facts.files.Block`, that already returns `list[str]`

### CODE REVIEW

**Comment by stone-w4tch3r** [comment]
#1028 
Removed dollar signs, looks like they were typo
#984 
As [mentioned](https://github.com/pyinfra-dev/pyinfra/issues/984#issuecomment-1565657899), renamed `files.marked_block` in documentation to `file.block`
#1031 
Adjusted `awk` to prevent duplications
#1029 
Added `try_prevent_shell_expansion` parameter to partially help with mentioned issue

And also changed type for `content` to `str | list[str]`. This makes it more unified with `facts.files.Block`, that already returns `list[str]`

File: pyinfra/operations/files.py [hunk 1]
```diff
@@ -1661,7 +1666,7 @@ def block(
         )
 
         # have two entries in /etc/host
-        files.marked_block(
+        files.block(
```
Comment (stone-w4tch3r) [inline pyinfra/operations/files.py:1670]: #984 
[this comment](https://github.com/pyinfra-dev/pyinfra/issues/984#issuecomment-1565657899)

File: pyinfra/operations/files.py [hunk 2]
```diff
@@ -1614,6 +1614,7 @@ def block(
     line=None,
     backup=False,
     escape_regex_characters=False,
+    try_prevent_shell_expansion=False,
```
Comment (stone-w4tch3r) [inline pyinfra/operations/files.py:1618]: #1029 
helps with mentioned issue

File: pyinfra/operations/files.py [hunk 3]
```diff
@@ -1748,21 +1773,30 @@ def block(
                 f"{print_after} f!=1 && /{regex}/ {{ print x; f=1}} "
                 f"END {{if (f==0) print ARGV[2] }} {print_before}'"
             )
-            cmd = StringCommand(out_prep, prog, q_path, f'$"{the_block}"', "> $OUT &&", real_out)
```
Comment (stone-w4tch3r) [inline pyinfra/operations/files.py:1751]: ...`f'$"{the_block}"'`...
Removed `$`

File: pyinfra/operations/files.py [hunk 4]
```diff
@@ -1748,21 +1773,30 @@ def block(
...
                 prog = (
                     'awk \'BEGIN {{f=1; x=ARGV[2]; ARGV[2]=""}}'
-                    f"/{mark_1}/ {{print; print x; f=0}} /{mark_2}/ {{print; f=1}} f'"
+                    f"/{mark_1}/ {{print; print x; f=0}} /{mark_2}/ {{print; f=1; next}} f'"
                 )
                 cmd = StringCommand(
                     out_prep,
                     prog,
                     q_path,
-                    '$"' + content + '"',
```
Comment (stone-w4tch3r) [inline pyinfra/operations/files.py:1765]: `'$"' + content + '"',`
Removed `$`

File: pyinfra/operations/files.py [hunk 5]
```diff
@@ -1748,21 +1773,30 @@ def block(
...
-            pieces = content.split("\n")
-            if (len(current) != len(pieces)) or (
-                not all(lines[0] == lines[1] for lines in zip(pieces, current))
+            if (len(current) != len(content)) or (
+                    not all(lines[0] == lines[1] for lines in zip(content, current))
             ):  # marked_block found but text is different
                 prog = (
                     'awk \'BEGIN {{f=1; x=ARGV[2]; ARGV[2]=""}}'
-                    f"/{mark_1}/ {{print; print x; f=0}} /{mark_2}/ {{print; f=1}} f'"
+                    f"/{mark_1}/ {{print; print x; f=0}} /{mark_2}/ {{print; f=1; next}} f'"
```
Comment (stone-w4tch3r) [inline pyinfra/operations/files.py:1793]: #1031 
Fixes this issue

**Review by Fizzadar — APPROVED** [review]
@stone-w4tch3r huge thank you for working on this, I have to admit I have been struggling greatly to wrap my head around this operation, hugely appreciate your work here. Finally have pushed up some minor changes and fixed the tests to get this in.

**Comment by Fizzadar** [comment]
(force push to get github to actually run actions, being very flakey at the moment)

**Pull request merged**